src/hypothesis/internal/conjecture/engine.py

```diff
@@ -1845,6 +1846,57 @@ def interval_deletion_with_block_lowering(self):
             if not changed:
                 i += 1
 
+    def lower_dependent_block_pairs(self):
+        """This is a fairly specific shrink pass that is mostly specialised for
+        our integers strategy, though is probably useful in other places.
+
+        It looks for adjacent pairs of blocks where lowering the value of the
+        first changes the size of the second. Where this happens, we may lose
+        interestingness because this takes the prefix rather than the suffix
+        of the next block, so if lowering the block produces an uninteresting
+        value and this change happens, we try replacing the second block with
+        its suffix and shrink again.
+
+        For example suppose we had:
+
+            m = data.draw_bits(8)
+            n = data.draw_bits(m)
+
+        And initially we draw m = 9, n = 1. This gives us the bytes [9, 0, 1].
+        If we lower 9 to 8 then we now read [9, 0], because the block size of
+        the n has changed. This pass allows us to also try [9, 1], which
+        corresponds to m=8, n=1 as desired.
+
+        This should *mostly* be handled by the minimize_individual_blocks pass,
+        but that won't always work because its length heuristic can be wrong if
+        the changes to the next block have knock on size changes, while this
+        one triggers more reliably.
+
+        """
+        self.debug('Lowering adjacent pairs of dependent blocks')
+        self.__engine.debug_data(self.shrink_target)
+        i = 0
+        while i + 1 < len(self.shrink_target.blocks):
+            u, v = self.shrink_target.blocks[i]
+            i += 1
+
+            b = int_from_bytes(self.shrink_target.buffer[u:v])
+            if b > 0:
+                attempt = bytearray(self.shrink_target.buffer)
+                attempt[u:v] = int_to_bytes(b - 1, v - u)
+                attempt = hbytes(attempt)
+                shrunk = self.cached_test_function(attempt)
+                self.__engine.debug_data(shrunk)
```

Comment:
Performance note: `ConjectureEngine.debug_data` has a significant performance cost even with debug logging disabled.  I think this is actually something I got wrong in #1042 - the conditional should be moved into the method, and do an early return there.  We're obviously inviting problems otherwise!

Reply:
Oh that was a debugging line that I'd intended to remove anyway. Whoops, thanks.

Reply:
I think it's still worth moving the check into the method - I simply didn't consider that more calls could be added elsewhere, which in hindsight is pretty obvious.  *Not* moving it is just asking for a needless performance hit when we miss one in review someday.

Reply:
I agree. I've added it in here because hey this pull request is already huge so might as well. 😉